### DISCUSSION TITLE
Feature request: In-house pre-commit hook integration

### DISCUSSION THREAD

**Discussion by MatthewScholefield** [astral-sh/rye#1095] [Ideas]
Since `rye` offers `rye lint --fix` and `rye fmt` out of the box, it's natural that people may want to run this as a pre-commit hook. There already exists [`pre-commit`](https://github.com/pre-commit/pre-commit) which has its own dependency management system which seems redundant in the context of `rye` (but for reference, [here's the mini config I use for it](https://github.com/MatthewScholefield/patchpy/blob/main/.pre-commit-config.yaml)). So I'm wondering if there is the potential to offer some simple in-house solution for this.

Right now it actually would work fine to just make some `.hooks/pre-commit` file which specifies `rye lint --fix && rye fmt` but this won't handle re-adding changed files and also won't have any nice simple UI to show which hooks are running, passed, etc.

Some potential implementations:
 - Have a `rye pre-commit install` and `rye pre-commit check`. The `check` command would read from `pyproject.toml` to determine the list of checks / scripts to run.
 - Do the same thing but outside of rye as some third party `rye-pre-commit` command (repo name something like *Rye integrated pre-commit hooks*)
 
 Overall, I'm not sure this is something others would want but I just wanted to bring it up and see what others think.

**Comment by manugarri** [comment]
+1 to this. Rye looks very promising, but using pre-commit allows developers to easily include a wide variety of hooks, and also ensure that every commit automatically passes qa checks without any manual intervention.

**Comment by vtnate** [comment]
Ruff has pre-commit integration, which is what `rye lint/fmt` are doing. https://github.com/astral-sh/ruff-pre-commit
